builder/bundle_control_test.go

```diff
@@ -0,0 +1,432 @@
+package builder
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	"github.com/clearlinux/mixer-tools/config"
+)
+
+func TestGetUpstreamBundlesVerDir(t *testing.T) {
+	testCases := []struct {
+		ver string
+		exp string
+	}{
+		{"0", "clr-bundles-0"},
+		{"000001", "clr-bundles-000001"},
+		{"non-numeric", "clr-bundles-non-numeric"},
+		{"", "clr-bundles-"},
+		{"25660", "clr-bundles-25660"},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.ver, func(t *testing.T) {
+			actual := getUpstreamBundlesVerDir(tc.ver)
+			if actual != tc.exp {
+				t.Errorf("expected %s on input %s but got %s", tc.exp, tc.ver, actual)
+			}
+		})
+	}
+}
+
+func TestGetUpstreamBundlesPath(t *testing.T) {
+	testCases := []struct {
+		ver string
+		exp string
+	}{
+		{"0", "test/upstream-bundles/clr-bundles-0/bundles"},
+		{"000001", "test/upstream-bundles/clr-bundles-000001/bundles"},
+		{"non-numeric", "test/upstream-bundles/clr-bundles-non-numeric/bundles"},
+		{"", "test/upstream-bundles/clr-bundles-/bundles"},
+		{"25660", "test/upstream-bundles/clr-bundles-25660/bundles"},
+	}
+	b := New()
+	b.Config = config.MixConfig{}
+	b.Config.LoadDefaultsForPath(false, "test")
+	b.Config.Builder.VersionPath = "test"
+	for _, tc := range testCases {
+		t.Run(tc.ver, func(t *testing.T) {
+			actual := b.getUpstreamBundlesPath(tc.ver)
+			if actual != tc.exp {
+				t.Errorf("expected %s on input %s but got %s", tc.exp, tc.ver, actual)
+			}
+		})
+	}
+}
+
+func TestGetLocalPackagesPath(t *testing.T) {
+	b := New()
+	// internally this function requires some sub-objects of builder to
+	// be set (non-nil)
+	b.Config = config.MixConfig{}
+	b.Config.LoadDefaultsForPath(false, "test")
+	b.Config.Builder.VersionPath = "test"
+	b.LocalPackagesFile = "local-packages"
+	actual := b.getLocalPackagesPath()
+	if actual != "test/local-packages" {
+		t.Errorf("expected test/local-packages but got %s", actual)
+	}
+}
+
+func TestGetUpstreamPackagesPath(t *testing.T) {
+	testCases := []struct {
+		ver string
+		exp string
+	}{
+		{"0", "upstream-bundles/clr-bundles-0/packages"},
+		{"000001", "upstream-bundles/clr-bundles-000001/packages"},
+		{"non-numeric", "upstream-bundles/clr-bundles-non-numeric/packages"},
+		{"", "upstream-bundles/clr-bundles-/packages"},
+		{"25660", "upstream-bundles/clr-bundles-25660/packages"},
+	}
+
+	b := New()
+	for _, tc := range testCases {
+		t.Run(tc.ver, func(t *testing.T) {
+			b.UpstreamVer = tc.ver
+			actual := b.getUpstreamPackagesPath()
+			if actual != tc.exp {
+				t.Errorf("expected %s but got %s", tc.exp, actual)
+			}
+		})
+	}
+}
+
+func TestGetUpstreamBundles(t *testing.T) {
+	b := New()
+	Offline = true
+	if b.getUpstreamBundles("", false) != nil {
+		t.Error("returned error when in offline mode")
+	}
+	// TODO: mock network
```

Comment:
Maybe we can mock this with setting the env at beginning of test with http_proxy=0.0.0.0 like we do in BATS, it should set it for the whole run and just work maybe? Since we don't care about slow network tests (right now) just lack of network.

Reply:
No there are actual mock packages out there that mock network responses.

That also doesn't "mock network". The TODO is so we can actually test the rest of `getUpstreamBundles`. It already tests offline mode.